### PR TITLE
Ben/t5

### PR DESCRIPTION
**Pull request by benleetownsend**

### CODE REVIEW

File: finetune/encoding/input_encoder.py [hunk 1]
```diff
@@ -203,11 +211,11 @@ def encode_multi_input(self, Xs, max_length=None, remove_repeated_whitespace=Fal
...
-        token_ids = self._cut_and_concat(encoded=encoded.token_ids, max_length=max_length)
-        tokens = self._cut_and_concat(encoded=encoded.tokens, max_length=max_length)
-        token_ends = self._cut_and_concat(encoded=encoded.token_ends, max_length=max_length, special_tokens=-1)
-        token_starts = self._cut_and_concat(encoded=encoded.token_starts, max_length=max_length, special_tokens=-1)
-
+        token_ids = self._cut_and_concat(encoded=encoded.token_ids, max_length=max_length, include_bos_eos=include_bos_eos)
+        tokens = self._cut_and_concat(encoded=encoded.tokens, max_length=max_length, include_bos_eos=include_bos_eos)
+        token_ends = self._cut_and_concat(encoded=encoded.token_ends, max_length=max_length, special_tokens=-1, include_bos_eos=include_bos_eos)
+        token_starts = self._cut_and_concat(encoded=encoded.token_starts, max_length=max_length, special_tokens=-1, include_bos_eos=include_bos_eos)
+        print("Finetune", token_ids)
```
Comment (benleetownsend) [inline finetune/encoding/input_encoder.py]: :-1:

File: finetune/target_models/seq2seq.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
+
+import tensorflow as tf
+
+from finetune.base import BaseModel
+from finetune.encoding.target_encoders import Seq2SeqLabelEncoder
+from finetune.input_pipeline import BasePipeline
+from finetune.util.shapes import shape_list
+from finetune.nn.target_blocks import language_model
```
Comment (benleetownsend) [inline finetune/target_models/seq2seq.py]: Unused imports

File: finetune/target_models/seq2seq.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+                    None, #output_attentions,
+                    None, #output_hidden_states,
+                ),
+                training=train,
+            )[0]
+            logits = featurizer_state["embed_weights"](normalize_embeds(embeds), mode="linear")
+            loss = tf.nn.sparse_softmax_cross_entropy_with_logits(
+                logits=logits, labels=targets[:, 1:], 
+            )
+            padding_mask = tf.compat.v1.Print(padding_mask, [padding_mask], summarize=100)
```
Comment (benleetownsend) [inline finetune/target_models/seq2seq.py]: Print

File: finetune/target_models/seq2seq.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+            return {
+                "logits": logits,
+                "losses": loss
+            }
+
+        else:
+            def symbols_to_logits_fn(input_symbols, i, state): #[batch_size, decoded_ids] to [batch_size, vocab_size]
+                with tf.compat.v1.variable_scope("model"):
+                    with tf.compat.v1.variable_scope("target"):
+                        input_symbols = tf.compat.v1.Print(input_symbols, [input_symbols])
```
Comment (benleetownsend) [inline finetune/target_models/seq2seq.py]: Print

File: finetune/target_models/seq2seq.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+                                None, #decoder_past_key_value_states,
+                                False, #use_cache,
+                                None, #output_attentions,
+                                None, #output_hidden_states,
+                            ),
+                            training=False,
+                        )[0]
+                        logits = featurizer_state["embed_weights"](normalize_embeds(embeds[:, -1]), mode="linear")
+                        logits_shape = tf.shape(logits)
+                        #logits = tf.concat((tf.zeros(shape=[logits_shape[0], 1], dtype=tf.float32), logits[:, 1:]), 1)
```
Comment (benleetownsend) [inline finetune/target_models/seq2seq.py]: Commented code

File: finetune/target_models/seq2seq.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+    def _predict_op(self, logits):
+        return logits
+
+    def _predict_proba_op(self, logits):
+        return logits
+
+    def _predict(self, zipped_data, **kwargs):
+        preds = self._inference(zipped_data, predict_keys=[PredictMode.NORMAL],  **kwargs)
+        print(preds)
```
Comment (benleetownsend) [inline finetune/target_models/seq2seq.py]: Print

File: finetune/util/huggingface_interface.py [hunk 7]
```diff
@@ -44,12 +45,13 @@ def finetune_model_from_huggingface(
...
     hf_config,
     weights_replacement,
+    include_bos_eos=True,
 ):
     weights_url = archive_map[pretrained_weights]
     hf_tokenizer_instance = hf_tokenizer.from_pretrained(pretrained_weights)
     hf_config_instance = hf_config.from_pretrained(pretrained_weights)
 
-    def finetune_featurizer(X, encoder, config, train=False, reuse=None, **kwargs):
+    def finetune_featurizer(X, encoder, config, train=False, reuse=None, targets=None, lengths=None, **kwargs):
```
Comment (benleetownsend) [inline finetune/util/huggingface_interface.py]: Should we use this on other models? I definite wrote this code once before but it seems it never got merged

File: finetune/util/huggingface_interface.py [hunk 8]
```diff
@@ -153,8 +167,16 @@ def _encode(self, texts):
                     batch_char_ends.append(token_ends)
                     batch_char_starts.append(token_starts)
                 else:
-                    if not hasattr(self.tokenizer, 'sp_model'):
-                        LOGGER.warning("Tokenizer is not sentence-piece-based and is not guaranteed to port over correctly.")
+                    if not hasattr(self.tokenizer, "sp_model"):
+                        LOGGER.warning(
+                            "Tokenizer is not sentence-piece-based and is not guaranteed to port over correctly."
+                        )
+                    # This may break some downstream finetune assumptions
```
Comment (benleetownsend) [inline finetune/util/huggingface_interface.py:176]: I think this is fine.

File: finetune/util/huggingface_interface.py [hunk 9]
```diff
@@ -178,16 +205,16 @@ def _encode(self, texts):
                     batch_char_ends.append(tok_pos)
                     batch_char_starts.append(char_starts)
 
-            print(batch_tokens, batch_char_starts)
-            return EncodedOutput(
+            output = EncodedOutput(
```
Comment (benleetownsend) [inline finetune/util/huggingface_interface.py]: Debugging change not reverted

File: tests/test_huggingface.py [hunk 10]
```diff
@@ -15,19 +29,32 @@ def setUp(self):
...
+            finetune_seq_features, hf_seq_features, decimal=5,
         )
+        finetune_model.fit([self.text], [[{"start": 0, "end": 4, "label": "class_a"}]])
 
     def huggingface_embedding(self, text, model_path):
         tokenizer = AutoTokenizer.from_pretrained(model_path)
         model = TFAutoModel.from_pretrained(model_path)
         input_ids = tf.constant(tokenizer.encode(self.text))[None, :]  # Batch size 1
-        outputs = model(input_ids)  # outputs is tuple where first element is sequence features
+        print(input_ids)
```
Comment (benleetownsend) [inline tests/test_huggingface.py]: Print

File: finetune/encoding/target_encoders.py [hunk 11]
```diff
@@ -124,13 +124,8 @@ def fit_transform(self, y):
...
         output = []
         for y_i in y:
-            out = self.encoder.encode_multi_input([[y_i]], max_length=self.max_len).token_ids
-            seq_length = len(out)
-            x = np.zeros((self.max_len, 2), dtype=np.int32)
-
-            x[:seq_length, 0] = out
-            x[:, 1] = np.arange(self.encoder.vocab_size, self.encoder.vocab_size + self.max_len)
-            output.append(x)
+            out = self.encoder.encode_multi_input([y_i], max_length=self.max_len, include_bos_eos=True).token_ids
```
Comment (madisonmay) [inline finetune/encoding/target_encoders.py:131]: Haha, this was already mostly done!  Nice

File: finetune/input_pipeline.py [hunk 12]
```diff
@@ -397,9 +418,9 @@ def _text_to_ids(self, Xs, pad_token=None):
...
 
             yield EncodedOutput(**d)
-    
+
     def __getstate__(self):
         state = self.__dict__.copy()
         del state["_text_encoder"]
         return state
-        
+
```
Comment (madisonmay) [inline finetune/input_pipeline.py:426]: Slowly we will black-ify all of our files :)

File: finetune/util/huggingface_interface.py (hunk 7 above)
Comment (madisonmay) [reply finetune/util/huggingface_interface.py]: Targets doesn't seem to be required any longer.  Do you mean the lengths argument?

File: finetune/model.py [hunk 13]
```diff
@@ -179,6 +179,8 @@ def _model_fn(features, labels, mode, params):
                 explain=build_explain,
                 context=context,
                 total_num_steps=total_num_steps,
+                targets=Y,
```
Comment (madisonmay) [inline finetune/model.py]: I don't think we need to pass targets to the featurizer any longer

File: finetune/util/huggingface_interface.py (hunk 7 above)
Comment (benleetownsend) [reply finetune/util/huggingface_interface.py]: Yep, lengths! I really meant to put this comment on the block that generates the lengths

File: finetune/target_models/seq2seq.py [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+        )
+
+    def _target_encoder(self):
+        return Seq2SeqLabelEncoder(self.text_encoder, self.config.max_length)
+
+def normalize_embeds(embeds):
+    return embeds * (tf.cast(tf.shape(embeds)[-1], dtype=tf.float32) ** -0.5)
+
+class HFS2S(BaseModel):
+
```
Comment (madisonmay) [inline finetune/target_models/seq2seq.py:41]: Future work -- exposing option to provide inputs to decoder that are conditioned on (at train or predict time) but act as a prefix / aren't part of the loss.

File: finetune/util/huggingface_interface.py (hunk 9 above)
Comment (madisonmay) [reply finetune/util/huggingface_interface.py]: Yup, my bad here.

File: tests/test_huggingface.py [hunk 15]
```diff
@@ -15,19 +29,32 @@ def setUp(self):
...
+
+        if model.config.is_encoder_decoder:
```
Comment (madisonmay) [inline tests/test_huggingface.py:50]: Can remove this conditional now

**Comment by madisonmay** [comment]
You somehow managed to make this so much cleaner than I expected.  Awesome stuff -- I'm excited to take this for a spin :D

File: finetune/util/huggingface_interface.py (hunk 7 above)
Comment (madisonmay) [reply finetune/util/huggingface_interface.py]: Yeah I'm down for that change.

File: tests/test_huggingface.py [hunk 16]
```diff
@@ -36,3 +61,45 @@ def test_xlm_roberta(self):
...
+        train_texts, test_texts, train_annotations, test_annotations = train_test_split(
+            texts, labels, test_size=0.1, random_state=42
+        )
+        finetune_model = HFS2S(
+            base_model=HFT5,
+            n_epochs=3,
+            batch_size=2,
+        )
+        finetune_model.fit(train_texts, train_annotations)
+        for t, p in zip(test_texts, finetune_model.predict(test_texts)):
```
Comment (madisonmay) [inline tests/test_huggingface.py]: Could we add a weak sanity check on the output here?
